In [6]:
# IMPORTS
import numpy as np
import pandas as pd

#Fin Data Sources
import yfinance as yf
import pandas_datareader as pdr

#Data viz
import plotly.graph_objs as go
import plotly.express as px

import time
from datetime import date

In [7]:
end = date.today()
print(f'Year = {end.year}; month= {end.month}; day={end.day}')

start = date(year=end.year-70, month=end.month, day=end.day)
print(f'Period for indexes: {start} to {end} ')

Year = 2024; month= 4; day=22
Period for indexes: 1954-04-22 to 2024-04-22 


In [8]:
# Real Potential Gross Domestic Product (GDPPOT), Billions of Chained 2012 Dollars, QUARTERLY
# https://fred.stlouisfed.org/series/GDPPOT
#gdppot = pdr.DataReader("GDPPOT", "fred", start=start)

# Real Gross Domestic Product (GDPC1), Billions of Chained 2017 Dollars, QUARTERLY
# https://fred.stlouisfed.org/series/GDPC1
gdpc1 = pdr.DataReader("GDPC1", "fred", start=start)

In [9]:
gdpc1['gdpc1_us_yoy'] = gdpc1.GDPC1/gdpc1.GDPC1.shift(4)-1
gdpc1['gdpc1_us_qoq'] = gdpc1.GDPC1/gdpc1.GDPC1.shift(4)-1
gdpc1last4 = gdpc1.tail(4)
avg_growth_gdp_2023 = gdpc1last4['gdpc1_us_yoy'].mean()
print(avg_growth_gdp_2023)

0.02540443392867625


[Question 1] 2.5%

In [10]:
# https://fred.stlouisfed.org/series/DGS1
#dgs1 = pdr.DataReader("DGS1", "fred", start=start)
#dgs1.tail()
dgs2 = pd.read_csv('/content/sample_data/DGS2.csv')
#dgs2
dgs10 = pd.read_csv('/content/sample_data/DGS10.csv')
dgs2j10 = dgs2.set_index('DATE').join(dgs10.set_index('DATE'))
dgs2j10['diff'] = dgs2j10['DGS10'].replace('.', 0).astype(float) - dgs2j10['DGS2'].replace('.', 0).astype(float)
min_value = dgs2j10['diff'].min()
print(min_value)

-1.0800000000000005


[Question 2] -1.1

In [11]:
# INDEXES from Yahoo Finance
# S&P 500 index
# WEB: https://finance.yahoo.com/quote/%5EGSPC/
sp500_daily = yf.download(tickers = "^GSPC",
                     period = "10y",
                     interval = "1d")

[*********************100%%**********************]  1 of 1 completed


In [12]:
sp500_daily.tail()

Open         High          Low        Close    Adj Close  \
Date                                                                          
2024-04-15  5149.669922  5168.430176  5052.470215  5061.819824  5061.819824   
2024-04-16  5064.589844  5079.839844  5039.830078  5051.410156  5051.410156   
2024-04-17  5068.970215  5077.959961  5007.250000  5022.209961  5022.209961   
2024-04-18  5031.520020  5056.660156  5001.890137  5011.120117  5011.120117   
2024-04-19  5005.439941  5019.020020  4953.560059  4967.229980  4967.229980   

                Volume  
Date                    
2024-04-15  3950210000  
2024-04-16  4006200000  
2024-04-17  3596130000  
2024-04-18  3619760000  
2024-04-19  3878750000

In [13]:
# prompt: get adj close on 2019-04-09

adj_close_2019_04_09_sp500 = sp500_daily.loc['2019-04-09']['Adj Close']
print(adj_close_2019_04_09_sp500)

2878.199951171875


In [14]:
# prompt: get adj close on 2024-04-09

adj_close_2024_04_09_sp500 = sp500_daily.loc['2024-04-09']['Adj Close']
print(adj_close_2024_04_09_sp500)


5209.91015625


In [15]:
five_year_growth_sp500_daily = (adj_close_2024_04_09_sp500-adj_close_2019_04_09_sp500)/adj_close_2019_04_09_sp500
print(five_year_growth_sp500_daily)

0.8101279426847173


In [16]:
# INDEXES from Yahoo Finance
# IPC MEXICO index
# WEB: https://finance.yahoo.com/quote/%5EMXX/
ipc_mx_daily = yf.download(tickers = "^MXX",
                     period = "10y",
                     interval = "1d")

[*********************100%%**********************]  1 of 1 completed


In [17]:
ipc_mx_daily.tail()

Open          High           Low         Close  \
Date                                                                 
2024-04-15  56536.468750  56709.121094  55924.101562  55984.000000   
2024-04-16  56009.730469  56148.000000  55626.750000  55797.250000   
2024-04-17  55788.738281  56090.320312  55383.250000  55415.691406   
2024-04-18  55597.789062  55941.621094  55585.820312  55739.429688   
2024-04-19  55811.191406  56036.718750  55579.691406  55862.851562   

               Adj Close     Volume  
Date                                 
2024-04-15  55984.000000  170645700  
2024-04-16  55797.250000  225862900  
2024-04-17  55415.691406  244258900  
2024-04-18  55739.429688  208671900  
2024-04-19  55862.851562  138617600

[Question 3] S&P 81%

In [18]:
#Question 4
q4_dct = {'2222.SR':None, 'BRK-B':None, 'AAPL':None, 'MSFT':None, 'GOOG':None, 'JPM':None}

for x in q4_dct.keys():
  q4_daily = yf.download(tickers = x,
                      start="2023-01-01",
                      end="2023-12-31",
                      interval = "1d")
  max_adj_close_q4_daily = q4_daily['Adj Close'].max()
  min_adj_close_q4_daily = q4_daily['Adj Close'].min()
  lrr_adj_close_q4_daily = (max_adj_close_q4_daily-min_adj_close_q4_daily)/max_adj_close_q4_daily
  q4_dct[x] = lrr_adj_close_q4_daily

q4_max_key = max(q4_dct, key=q4_dct.get)
print("\n", q4_max_key, "has the largest range ratio at", round(q4_dct.get(q4_max_key), 2))

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


 MSFT has the largest range ratio at 0.42


[Question 4] MSFT 0.42

In [19]:
#refactor - multiple tickers in one set
q4_2_dct = {'2222.SR':None, 'BRK-B':None, 'AAPL':None, 'MSFT':None, 'GOOG':None, 'JPM':None}
q4_2_daily = yf.download(tickers = list(q4_2_dct.keys()),
                      start="2023-01-01",
                      end="2023-12-31",
                      interval = "1d")
#q4_2_daily.tail()

for x in q4_2_dct.keys():
  max_adj_close_q4_2_daily = q4_2_daily['Adj Close'][x].max()
  min_adj_close_q4_2_daily = q4_2_daily['Adj Close'][x].min()
  lrr_adj_close_q4_2_daily = (max_adj_close_q4_2_daily-min_adj_close_q4_2_daily)/max_adj_close_q4_2_daily
  q4_2_dct[x] = lrr_adj_close_q4_2_daily

q4_2_max_key = max(q4_2_dct, key=q4_2_dct.get)
print("\n", q4_2_max_key, "has the largest range ratio at", round(q4_2_dct.get(q4_2_max_key), 2))

[*********************100%%**********************]  6 of 6 completed


 MSFT has the largest range ratio at 0.42


In [20]:
#Question 5
q5_dct = {'2222.SR':None, 'BRK-B':None, 'AAPL':None, 'MSFT':None, 'GOOG':None, 'JPM':None}

for x in q5_dct.keys():
  q5_ticker = yf.Ticker(x)
  q5_dividends_2023 = q5_ticker.dividends[q5_ticker.dividends.index.year == 2023].sum()
  q5_daily = yf.download(tickers = x,
                      start="2023-01-01",
                      end="2023-12-31",
                      interval = "1d")
  q5_adj_close_eoy_2023 = q5_daily.loc[q5_daily.index[-1]]['Adj Close']
  q5_yield_2023 = q5_dividends_2023/q5_adj_close_eoy_2023
  q5_dct[x] = q5_yield_2023

q5_max_key = max(q5_dct, key=q5_dct.get)
print("\n", q5_max_key, "has the largest yield value at", round(q5_dct.get(q5_max_key)*100, 1), "%")

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


 2222.SR has the largest yield value at 2.8 %


[Question 5] 2222.SR dividend yield is 2.8%

[Question 6]

Currently considering project to be about US tire market.

GDP for countries representing the market
*   FRED - US Potential GDP (GDPOT) - economic growth; check for consistency of supply impact on price;...but if you have a vehicle, you need tires
*   The World Bank - GDP of other countries will big players like France and Japan

Yahoo Finance growth/CAGR/OHLCV/earnings data for publicly-traded companies in the market
*   Major tire sellers like Bridgestone, Michelin, Goodyear;...but tires go on vehicles
*   Major vehicle retailers like Tesla, Toyota, GM, and Ford
*   Major vehicle resellers like Carvana, CarMax, and AutoNation

Need to look into where to get performance info on private companies like TireRack, DiscountTire, and SimpleTire

Need to look into where to get info on global markets of rubber




[Question 7]
If not available in the yfinance module, go for a web scraping approach.
- Scrape the grid for multiple day values using the day parameter in the URL's querystring
- Iterate the Reported EPS for a year of data
- Maintain a list of companies you are interested in
- Create a Pandas DataFrame with Estimated EPS, Reported EPS, and Surprise % as columns and quarters as the index
- Scrape the grid being returned by the web page, and record into the dataframe